src/server/accounting.c

```diff
@@ -2201,3 +2222,91 @@ account_job_update(job *pjob, int type)
 		free(resc_used);
 	}
 }
+
+/**
+ * @brief
+ * 	log an alter record for modified jobs
+ * 	plist contains the attributes and resources requested to be modified.
+ * 	We only modify those because the ATTR_l encode function will encode
+ * 	all resources, not just the ones we want.
+ * 
+ * @param[in] pjob - job to log records for.
+ * @param[in] plist - list of attributes and resources to log
+ * 
+ * @returns void 
+ */
+void log_alter_records_for_attrs(job *pjob, svrattrl *plist) {
+	svrattrl *svrattrl_list;
+	svrattrl *cur_svr;
+	pbs_list_head phead;
+	svrattrl *cur_plist;
+	char *per_alter_buf = NULL;
+	char *entire_record = NULL;
+	int entire_record_len = 0;
+
+	int i;
+	entire_record = malloc(1024 * sizeof(char));
+	if (entire_record == NULL)
+		return;
+	entire_record_len = 1024;
+	entire_record[0] = '\0';
+	int error = 0;
+
+	CLEAR_HEAD(phead);
+	for (i = 0; i < JOB_ATR_LAST; i++) {
+		if (pjob->ji_wattr[i].at_flags & ATR_VFLAG_MODIFY) {
+			job_attr_def[i].at_encode(&pjob->ji_wattr[i], &phead, job_attr_def[i].at_name, NULL, ATR_ENCODE_CLIENT, &svrattrl_list);
+			for (cur_plist = plist; cur_plist != NULL; cur_plist = (svrattrl *)GET_NEXT(cur_plist->al_link)) {
+				if (strcmp(cur_plist->al_name, job_attr_def[i].at_name) != 0)
+					continue;
+				else {
+					for (cur_svr = svrattrl_list; cur_svr != NULL; cur_svr = (svrattrl *)GET_NEXT(cur_svr->al_link)) {
+						if (pjob->ji_wattr[i].at_type == ATR_TYPE_RESC) {
+							if (cur_plist->al_resc != NULL) {
+								if (strcmp(cur_plist->al_resc, cur_svr->al_resc) == 0) {
+									char *fmt;
+									if (strchr(cur_svr->al_value, ' ') == NULL)
+										fmt = "%s.%s=%s";
```

Comment:
can you add a space in the beginning of the format so that you don't have to do pbs_strcat for adding a space later on? Of course, you might have to keep track of not having space when first key-value pair is added to log. I feel it will perform better that way.

Reply:
There isn't a really good way to do this.  I don't want the space in the for the first record.  This means I'd have to do an if/else for each pbs_asprintf() (or an ugly ? : ) case to either add the space or not.  It is highly unlikely the pbs_strcat() will cause an allocation.  I allocate 1024 bytes.  That'll be enough space for most log records.  I'd rather leave this alone.